File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] input= br.readLine().split(" ");

        int K = Integer.parseInt(input[0]);
        int N = Integer.parseInt(input[1]);

        int max = Integer.MIN_VALUE;

        int[] lan = new int[K];

        for (int i = 0; i < K; i++) {
            int lanLength = Integer.parseInt(br.readLine());
            if (lanLength > max) {
                max = lanLength;
            }
            lan[i] = lanLength;
        }

        Arrays.sort(lan);
        long start = 1;
        long end = max;
        while (start <= end) {
            long cnt = 0;
            long mid = (start + end) / 2;

            for (int i = 0; i < K; i++) {
                cnt += lan[i] / mid;
            }

            if (cnt >= N) {
                start = mid + 1;
            } else {
                end = mid - 1;
            }
        }

        System.out.println(end);
    }
}

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
import sys

n, m = map(int, sys.stdin.readline().split())
lans = [int(sys.stdin.readline()) for _ in range(n)]

start = 1
end = max(lans)

while start <= end:
    target = (start + end) // 2

    count = 0
    for lan in lans:
        count += lan // target

    if count >= m:
        start = target + 1
    else:
        end = target - 1

print(end)
